search/includes/classes/class-versioning.php

```diff
@@ -745,4 +755,128 @@ public function action__ep_delete_indexable( $object_id, $indexable_slug ) {
 		// Clear the flag to return to normal
 		$this->is_doing_object_delete = false;
 	}
+
+	/**
+	 * Check if the versions are persisted correctly. If not recreate them.
+	 */
+	public function maybe_self_heal() {
+		if ( $this->is_versioning_valid() ) {
+			return;
+		}
+
+		// $indicies = $this->get_all_accesible_indicies();
+		// $versioning = $this->reconstruct_versioning( $indicies );
+
+		// update_option( self::INDEX_VERSIONS_OPTION, $versioning );
+	}
+
+	public function is_versioning_valid() {
+		$all_versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+
+		if ( ! is_array( $all_versions ) || count( $all_versions ) == 0 ) {
+			return false;
+		}
+
+		$valid_slugs = \ElasticPress\Indexables::factory()->get_all( null, true );
+
+
+		foreach ( $all_versions as $slug => $versions ) {
+			if ( ! in_array( $slug, $valid_slugs ) ) {
+				return false;
+			}
+
+			if ( ! is_array( $versions ) || count( $versions ) == 0 ) {
+				return false;
+			}
+
+			$active_statuses = wp_list_pluck( $versions, 'active' );
+
+			$array_index = array_search( true, $active_statuses, true );
+
+			if ( false === $array_index ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function get_all_accesible_indicies() {
+		$found_indices = [];
+
+		$response = $this->elastic_search_instance->remote_request( '_cat/indices?format=json' );
+
+		$response_code = (int) wp_remote_retrieve_response_code( $response );
+
+		if ( is_wp_error( $response ) || $response_code >= 400 ) {
+			return $found_indices;
+		}
+
+		$response_body_json = wp_remote_retrieve_body( $response );
+		$response_body = json_decode( $response_body_json, true );
+
+		if ( ! is_array( $response_body ) ) {
+			return $found_indices;
+		}
+
+		foreach ( $response_body as $index_obj ) {
+			if ( is_array( $index_obj ) && isset( $index_obj['index'] ) ) {
+				$found_indices[] = $index_obj['index'];
+			}
+		}
+
+		return $found_indices;
+	}
+
+	public function reconstruct_versioning( $indicies ) {
```

Comment:
```suggestion
	public function reconstruct_versioning_option( $indicies ) {
```

Small tweak - I'd suggest naming this so that it's more clear what exactly is being reconstructed.

Reply:
Fixed